### PR TITLE
SASS-589 : Added duplicate name validation for interest account pages…

### PR DESCRIPTION
**Pull request by robert-webb-hmrc**
… + some test refactor

### Description
Added validation to Taxed and Untaxed interest pages to check cya data for existing names, and will error on an duplicate entry. Added some test refactoring.

Add a link to the relevant story in Jira
[SASS-589](https://jira.tools.tax.service.gov.uk/browse/SASS-589)

##### Before Reviewing
- [ ]  Have you pulled the branch down?
- [ ]  Have you assigned yourself to the PR?
- [ ]  Have you moved the task to “in review” on JIRA?
- [ ]  Have you checked to ensure all dependencies are up to date?

##### Whilst Reviewing
- [ ]  Have you run the tests?
- [ ]  Have you run the journey tests?
- [ ]  Have you looked at the JIRA story to make sure all Acceptance Criteria has been met?

##### After Reviewing
- [ ]  Have you checked for merge conflicts?
- [ ]  Have you checked to make sure there are no builds in the pipeline before you merge?
- [ ]  Have you moved the task to “in pipeline” on Jira?

##### Before creating PR
- [x]  Have you run the tests?
- [x]  Have you run the journey tests?
- [x]  Have you addressed warnings where appropriate?

##### After PRs been Reviewed
- [ ]  Have you checked the PR Builder passes?
- [ ]  Have you checked for merge conflicts?
- [ ]  Have you checked code coverage isn’t lower than previously?

### CODE REVIEW

File: app/forms/validation/StringConstraints.scala [hunk 1]
```diff
@@ -34,8 +34,12 @@ object StringConstraints {
     x => if (x.length <= maxChars) Valid else Invalid(msgKey)
   )
 
-  def validateNotDuplicate(previousEntrys: List[String]): String => Constraint[String] = msgKey => constraint[String](
-    x => if (previousEntrys.contains(x)) Invalid(msgKey) else Valid
+  def validateNotDuplicateGiftAid(previousEntries: List[String]): String => Constraint[String] = msgKey => constraint[String](
```
Comment (sraspin) [inline app/forms/validation/StringConstraints.scala:37]: Is there are reason for having duplicate methods here?

Comment (robert-webb-hmrc) [reply app/forms/validation/StringConstraints.scala:37]: The Account names are parsed differently between the gift aid and interests api calls, they're parsed as a list of accounts in gift aid but have to be mapped into a sequence for interest

File: app/controllers/interest/TaxedInterestAmountController.scala [hunk 2]
```diff
@@ -47,16 +47,19 @@ class TaxedInterestAmountController @Inject()(
...
   implicit val executionContext: ExecutionContext = mcc.executionContext
-  val taxedInterestAmountForm: Form[TaxedInterestModel] = TaxedInterestAmountForm.taxedInterestAmountForm()
 
   def show(taxYear: Int, id: String): Action[AnyContent] = commonPredicates(taxYear, INTEREST).apply { implicit user =>
 
     val optionalCyaData = getModelFromSession[InterestCYAModel](SessionValues.INTEREST_CYA)
 
     val idMatchesPreviouslySubmittedAccount: Boolean = optionalCyaData.flatMap(_.taxedUkAccounts.map(_.exists(_.id.contains(id)))).getOrElse(false)
 
+    val previousNames: Seq[String] = optionalCyaData.flatMap(_.taxedUkAccounts.map(_.map(_.accountName))).getOrElse(Seq(""))
```
Comment (BenCosfordHMRC) [inline app/controllers/interest/TaxedInterestAmountController.scala:57]: .getOrElse(Seq("")) can be .getOrElse(Seq()), it doesn't need an empty string in there

File: app/forms/validation/StringConstraints.scala (hunk 1 above)
Comment (robert-webb-hmrc) [reply app/forms/validation/StringConstraints.scala]: generalised to Seq

**Pull request merged**